File: backend/app/schemas/workspace.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.analysis import (
    AtsCheck,
    ComparisonResult,
    GrammarIssue,
    ScoreResult,
    SectionResult,
    SuggestionCard,
)
from app.schemas.anchors import SuggestionAnchorRecord


class WorkspaceFileInfo(BaseModel):

    filename: str | None = None
    mime_type: str | None = None
    size: int | None = None
    extension: str | None = None


class WorkspaceDocumentPayload(BaseModel):

    source_text: str | None = None
    sections: list[SectionResult] = Field(default_factory=list)
    draft_content: dict | None = None


class WorkspaceAnalysisContext(BaseModel):

    scores: ScoreResult | None = None
    ats_checks: list[AtsCheck] = Field(default_factory=list)
    suggestions: list[SuggestionCard] | None = None
    comparison_result: ComparisonResult | None = None
    comparison_status: str | None = None
    grammar_issues: list[GrammarIssue] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)


class ChatMessage(BaseModel):

    timestamp: str
    role: Literal["user", "assistant"]
    content: str
    status: Literal["complete", "streaming", "error"] = "complete"


class WorkspaceNavigation(BaseModel):

    workspace_url: str
    results_url: str


class WorkspaceHydration(BaseModel):

    job_id: str
    status: Literal["preparing", "ready", "failed"]
    file: WorkspaceFileInfo
    document: WorkspaceDocumentPayload
    analysis: WorkspaceAnalysisContext
    navigation: WorkspaceNavigation
    suggestion_anchors: list[SuggestionAnchorRecord] = Field(default_factory=list)
    messages: list[ChatMessage] = Field(default_factory=list)
    error: str | None = None


class WorkspaceContentPatch(BaseModel):

    sections: dict[str, Any]


class WorkspaceContentSaveResult(BaseModel):

    saved: bool
    updated_at: str


class WorkspaceFileUrl(BaseModel):

    file_url: str
    expires_in: int = 3600

File: frontend/lib/workspace.ts
import { apiFetch } from "@/lib/api";
import type { JSONContent } from "@tiptap/core";
import type {
  AtsCheck,
  ComparisonResult,
  ScoreResult,
  SectionResult,
  SuggestionCard,
  GrammarIssue,
} from "@/lib/types";
import type { ChatMessage } from "@/lib/stores/workspace-v2-store";

export interface AnchorRect {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface SuggestionAnchorRecord {
  suggestion_id: string;
  section: string;
  text_anchor: string;
  page_index: number;
  rect: AnchorRect;
  priority: "high_impact" | "quick_win";
}

export interface WorkspaceFileInfo {
  filename: string | null;
  mime_type: string | null;
  size?: number | null;
  extension?: string | null;
}

export interface WorkspaceDocumentPayload {
  source_text: string | null;
  sections: SectionResult[];
  draft_content?: Record<string, JSONContent> | null;
}

export interface WorkspaceAnalysisContext {
  scores: ScoreResult | null;
  ats_checks: AtsCheck[];
  suggestions?: SuggestionCard[] | null;
  skills?: string[];
  grammar_issues?: GrammarIssue[];
  comparison_result?: ComparisonResult | null;
  comparison_status?: "pending" | "comparing" | "complete" | "failed" | null;
}

export interface WorkspaceNavigation {
  workspace_url: string;
  results_url: string;
}

export interface WorkspaceHydration {
  job_id: string;
  status: "preparing" | "ready" | "failed";
  file: WorkspaceFileInfo;
  document: WorkspaceDocumentPayload;
  analysis: WorkspaceAnalysisContext;
  navigation: WorkspaceNavigation;
  error?: string | null;
  suggestion_anchors?: SuggestionAnchorRecord[];
  messages?: ChatMessage[];
}

export async function getWorkspaceHydration(
  jobId: string
): Promise<WorkspaceHydration> {
  return apiFetch<WorkspaceHydration>(`/jobs/${jobId}/workspace`);
}

export interface WorkspaceFileUrl {
  file_url: string;
  expires_in: number;
}

export async function getJobFileUrl(jobId: string): Promise<WorkspaceFileUrl> {
  return apiFetch<WorkspaceFileUrl>(`/jobs/${jobId}/file`);
}